### DISCUSSION TITLE
FAQs, HowTos, Documentation and Announcements

### DISCUSSION THREAD

**Discussion by 9rb** [rchain/rchain#3508] [Announcements]
We will use this to publish FAQs, HowTos, Documentation and Announcements.

**We opened up GitHub discussions to encourage developer discussions and mutual help.**

**Comment by mjvallone** [comment]
I'll share a compilation of the knowledge I could get in a couple of weeks working in a specific task related to RChat. Feel free to make suggestions :)
I worked in this [repo](https://github.com/rchain/rchat), so everything that could be "said" here is related to that repository or to general knowledge of RChain.


- `bonds.txt` contents list of public keys and stake. Delimiter is space. The pair of public key and stake in a line.
- `wallets.txt` contents list of REVs and balance over comma. One pair of REV and balance in a line.

- JS code creates deploy and pushes to the RNode. To update on-chain data these deploys must be executed and packaged in blocks.
Bonds and wallets files are used to initialize the creation of genesis block (when on-chain state is empty). Bonds contains initial validators (their public keys) with corresponding stake, and wallets contains initial REV addresses with corresponding balances.

- `explore-deploy` API, is a POST call containing Rholang code, and this API **returns you containment of the first name created in the code**.

- You can only retrieve messages for specific URI (DB_CONTRACT_URI) which was used in chain_replica script to insert data.

- To create a new Zulip instance (when you want to have more than one running at the same time): Docker-compose has context in the folder where is the config file, so you can clone the repo in another folder and start another instance. Because postgres port is bound to the host you need to change it.

- We want to transfer messages from the first Zulip instance to postgres DB of the second Zulip instance. RChain node is the bridge which is populated by the chain_replica script. Data is in on-chain contract registered on this URI IDDB_CONTRACT_URI. So the second Zulip instance only needs data to be written in its postgres DB and you don't need to run any RChain related script. Data in IDDB_CONTRACT_URI contract should be read and pushed directly to postgres DB on the second Zulip instance which will then show messages from the first Zulip instance.

- All RChain nodes that are part of the same shard will have the same state (after sync of blocks), so in case of Zulip configuration boot and read nodes contains the same state. E.g. boot node can be in Argentina where you insert data from your Zulip instance (into registered DB_CONTRACT_URI contract) and I can run my read node in Croatia. After sync I can read from my node DB_CONTRACT_URI contract and it will have all data you inserted on your boot node. So DB_CONTRACT_URI is like server on the internet accessible by multiple users but in case of RChain this is decentralized server where each of us runs its own copy.

- After you create fresh instance of on-chain DB with myzulipdb-deploy, registered contract will be empty (DB_CONTRACT_URI will be newly generated), so to add data new messages in Zulip must be created and chain_replica must be running to push these messages to this empty DB_CONTRACT_URI.


**Questions & Answers**
Q: if I do not "propose" a deploy, will it connect to the blockchain? i.e. do I have to "propose" in order to obtain the query result even if it is a query to the on-chain data?
A: yes, you have to "propose" in order to get a (non-exploratory) deploy to run.

Q: is it right to use db!("select", "zerver_message", @wanted, *dbCh)? what @wanted should be (besides a list)?
A: should be a list of records ids, IIRC

Q: when creating a new instance of zulip, I should run iddb-deploy and myzulipdb-deploy.. right? what IDDB_CONTRACT_URI and DB_CONTRACT_URI should I use? IDDB_CONTRACT_URI generated by iddb-deploy and DB_CONTRACT_URI from the existing zulip instance (the one with messages I want to get)?
A: On the destination Zulip instance you can only use postgres DB to send data read from RChain source DB stored in DB_CONTRACT_URI. So on the destination side you don't need to initialize on-chain part.

Q: how can I get existing messages from the whole blockchain? now I'm getting the one from DB_CONTRACT_URI (the last space created in the registry you've mentioned yesterday)
A: You can only retrieve messages for specific URI (DB_CONTRACT_URI) which was used in chain_replica script to insert data.
